### PR TITLE
[WIP] Allow partial signing of Threshold condition

### PR DESCRIPTION
**Pull request by TimDaub**
This is a work in progress.
See Comments for more information.

TODO:
- What to do when a `PreimageSha256` is encountered?

### CODE REVIEW

File: cryptoconditions/types/threshold_sha256.py [hunk 1]
```diff
@@ -50,6 +52,69 @@ def __init__(self, threshold=None):
...
+                # NOTE: Now we replace all unsigned sub conditions, with filler
+                #       conditions to allow for the serialization of the
+                #       ThresholdSha256Fulfillment.
+                for unsigned_subcond in unsigned_subconds:
+                    self.add_subcondition(unsigned_subcond['body'].condition)
+                    # NOTE: This might looks strange, but `add_subcondition`
+                    #       does some magic, so replacing the index at
+                    #       `self.subconditions` wouldn't work.
+                    self.subconditions.remove(unsigned_subcond)
+                # NOTE: We break form the signing, as we have fulfilled the
```
Comment (TimDaub) [inline cryptoconditions/types/threshold_sha256.py]: *from not *form

File: cryptoconditions/types/threshold_sha256.py [hunk 2]
```diff
@@ -562,8 +627,8 @@ def validate(self, message=None, **kwargs):
         #     # Fulfillment is not minimal
         #     return False
         # TODO: ILP specs see unfulfilled conditions as conditions and not fulfillments
-        valid_decisions = []
+        valid_decisions = 0
         for fulfillment in fulfillments:
             if fulfillment['body'].validate(message, **kwargs):
-                valid_decisions += [True] * fulfillment['weight']
```
Comment (TimDaub) [inline cryptoconditions/types/threshold_sha256.py]: :trollface:

**Comment by diminator** [comment]
I'm afraid that having a sign method in a ThresholdCondition is out of spec for the library.
- It won't appear in the JS, Java or other library. Hence yielding compatibility issues
- It's only relevant for subconditions that are Ed25519, but there is also RSA, and maybe other schemes

If you want to make this abstraction, it should be outside of the CC library

File: cryptoconditions/types/threshold_sha256.py (hunk 2 above)
Comment (diminator) [inline cryptoconditions/types/threshold_sha256.py]: ;-)

That's indeed a valid shortcut
PS: ILP/CC is probably removing weights in the near future

**Comment by diminator** [comment]
[Here](https://github.com/interledger/five-bells-condition) is the reference implementation

I'd want to keep the basic API as close to that as possible.

extra types such as `Timeout` and `InvertedThreshold` can only be used internally in BDB, as well as added functionality like you are proposing. I would propose to use this repo as a clean and in-sync repo. If we want to do custom stuff (including `Timeout` and `InvertedThreshold`) then we might want to embed that in our own repo through inheritance or so.

**Comment by TimDaub** [comment]
I've worked on it further.
Currently I don't see a problem in "supporting" `PreimageSha256`. It simply has no notion of `sign` which means this method will sign it neither. 
It gets trickier to know though when signing can be stopped as the `threshold` has been reached.
For now I'm using `self.validate` for that, but as `PreimageSha256` always returns `True` (https://github.com/bigchaindb/cryptoconditions/issues/26) knowing when to stop is kind of tricky (or even impossible at this point?).

**Comment by diminator** [comment]
> It gets trickier to know though when signing can be stopped as the threshold has been reached.

I wouldnt bother with that, kinda pre-optimization
just sign everything and validate later.

File: cryptoconditions/types/threshold_sha256.py [hunk 3]
```diff
@@ -50,6 +52,79 @@ def __init__(self, threshold=None):
...
+        This method will take the currently configured values for the message
+        prefix and suffix and create a signature using the provided
+        ThresholdSha256 private keys.
+
+        Args:
+            message (str): message to be signed
+            # TODO: Pretty sure that also other keys than Ed25519 keys can be
+                    used to sign, as a ThresholdSha256Fulfillment can not only
+                    consist of them but also other fulfillments.
+            private_keys (list(SigningKey)) Ed25519 private key
```
Comment (diminator) [inline cryptoconditions/types/threshold_sha256.py]: Practically you would almost never pass a list of signing keys unless you own them all

File: cryptoconditions/types/threshold_sha256.py [hunk 4]
```diff
@@ -50,6 +52,79 @@ def __init__(self, threshold=None):
...
+            else:
+                # TODO: Implement partial signing for other types.
+                raise NotImplementedError('Partial signing is only '
+                                          'supported for '
+                                          'Ed25519Fulfillments and '
+                                          'ThresholdSha256Fulfillments.')
+
+        # NOTE: If `self.validate` yields true, this means the threshold has
+        #       been reached.
+        if self.validate(message):
```
Comment (diminator) [inline cryptoconditions/types/threshold_sha256.py]: you can remove this - premature optimization

**Pull request closed without merging**